### PR TITLE
Feature/ops rand

### PR DESCRIPTION
**Pull request by Ashutosh-Londhe**
Added OPS_RAND API

### CODE REVIEW

File: apps/c/random/.generated
Comment (reguly) [inline apps/c/random/.generated:1]: don't commit this file

File: ops/c/src/core/ops_lib_core.cpp
Comment (reguly) [inline ops/c/src/core/ops_lib_core.cpp:1]: Use OPS_instance::OPS_internal_0 instead.

**Review by reguly — CHANGES_REQUESTED** [review]
The CPU-side is essentially replicated 3 times. It should only exist once in the backend - as a "host" implementation, which then marks the dat as dirty on the host, so if device applications subsequently want to read it, it will get uploaded. CPU backends just call this single implementation, and device backends (OpenCL, SYCL), should just call the host version behind the scenes.

File: ops/c/src/core/ops_lib_core.cpp
Comment (Ashutosh-Londhe) [reply ops/c/src/core/ops_lib_core.cpp:1]: done

File: ops/c/include/ops_internal2.h [hunk 1]
```diff
@@ -287,6 +287,20 @@ void ops_put_data(ops_dat dat);
 OPS_FTN_INTEROP
 void create_kerneldesc_and_enque(char const *name, ops_arg *args, int nargs, int index, int dim, int isdevice, int *range, ops_block block, void (*func)(struct ops_kernel_descriptor *desc));
 
+
+/*******************************************************************************
+* Random number generations
+*******************************************************************************/
+void ops_randomgen_init(unsigned int seed, int options);
```
Comment (reguly) [inline ops/c/include/ops_internal2.h:294]: These ones (not the host ones) should go to ops_lib_core.h

File: ops/c/include/ops_internal2.h [hunk 2]
```diff
@@ -287,6 +287,20 @@ void ops_put_data(ops_dat dat);
...
+void ops_fill_random_uniform(ops_dat dat);
+void ops_fill_random_normal(ops_dat dat);
+void ops_randomgen_exit();
+
+void ops_randomgen_init_host(unsigned int seed, int options, std::mt19937 ops_rand_gen);
```
Comment (reguly) [inline ops/c/include/ops_internal2.h:299]: What is the reason for passing ops_rand_gen through? Especially if it's fixed to having std::mt19937 as a type?

**Review by reguly — CHANGES_REQUESTED** [review]

File: ops/c/include/ops_internal2.h (hunk 2 above)
Comment (Ashutosh-Londhe) [reply ops/c/include/ops_internal2.h:299]: The host is common implementation among host, sycl and opencl

so in sycl_common, host_common and opencl_common we have ops_randomgen_init which will call this implementation at backend as this similar for all
thats why passing here the global delcared in those files here for initilization

File: ops/c/include/ops_internal2.h (hunk 1 above)
Comment (Ashutosh-Londhe) [reply ops/c/include/ops_internal2.h:294]: ok

File: ops/c/include/ops_internal2.h (hunk 2 above)
Comment (reguly) [reply ops/c/include/ops_internal2.h:294]: I understand why we call to ops_randomgen_init_host - my issue is why have the ops_rand_gen global three times in the three backends, when its only use is in ops_lib_core.cpp. Why not there?

Comment (Ashutosh-Londhe) [reply ops/c/include/ops_internal2.h:294]: yes as per your suggestion, i moved common implementation to ops_lib_core.cpp which this sycl, host and opencl will call

should i move the declarations also from ops_internal2.h to ops_lib_core.h

Comment (reguly) [reply ops/c/include/ops_internal2.h:294]: No, those are internal functions the use should not be calling directly, so they belong in ops_internal2.h

**Review by reguly — APPROVED** [review]

**Pull request merged**